File: labs/algorithms-live-coding/java/src/main/java/learning/livecoding/java/RetryBackoffExercise.java
package learning.livecoding.java;

import java.util.ArrayList;
import java.util.List;

public final class RetryBackoffExercise implements Exercise {
    @Override
    public String id() {
        return "retry-backoff";
    }

    @Override
    public String title() {
        return "Retry with Exponential Backoff";
    }

    @Override
    public String summary() {
        return "Retry only transient failures, increase the delay each time, and stop once the call succeeds or the retry budget is exhausted.";
    }

    record RetryResult(String value, int attempts, List<Long> scheduledBackoffsMs, String status) {
    }

    static final class TransientFailure extends RuntimeException {
        TransientFailure(String message) {
            super(message);
        }
    }

    static final class PermanentFailure extends RuntimeException {
        PermanentFailure(String message) {
            super(message);
        }
    }

    static final class FlakyPaymentGateway {
        private final int transientFailuresBeforeSuccess;
        private final boolean permanentFailure;
        private int calls;

        FlakyPaymentGateway(int transientFailuresBeforeSuccess, boolean permanentFailure) {
            this.transientFailuresBeforeSuccess = transientFailuresBeforeSuccess;
            this.permanentFailure = permanentFailure;
        }

        String capture() {
            calls++;
            if (permanentFailure) {
                throw new PermanentFailure("card-declined");
            }
            if (calls <= transientFailuresBeforeSuccess) {
                throw new TransientFailure("timeout");
            }
            return "captured";
        }
    }

    static RetryResult retryWithBackoff(int maxAttempts, long initialBackoffMs, ThrowingSupplier<String> operation) {
        int attempts = 0;
        long nextBackoffMs = initialBackoffMs;
        List<Long> scheduledBackoffs = new ArrayList<>();

        while (attempts < maxAttempts) {
            attempts++;
            try {
                return new RetryResult(operation.get(), attempts, List.copyOf(scheduledBackoffs), "success");
            } catch (PermanentFailure ignored) {
                return new RetryResult(null, attempts, List.copyOf(scheduledBackoffs), "permanent-failure");
            } catch (TransientFailure ignored) {
                if (attempts == maxAttempts) {
                    return new RetryResult(null, attempts, List.copyOf(scheduledBackoffs), "exhausted");
                }
                scheduledBackoffs.add(nextBackoffMs);
                nextBackoffMs *= 2;
            }
        }

        throw new IllegalStateException("Retry loop should have returned before reaching this point");
    }

    @Override
    public void run() {
        FlakyPaymentGateway successGateway = new FlakyPaymentGateway(2, false);
        ExerciseSupport.expectEquals(
            "eventual-success",
            new RetryResult("captured", 3, List.of(100L, 200L), "success"),
            retryWithBackoff(4, 100, successGateway::capture)
        );

        FlakyPaymentGateway permanentFailureGateway = new FlakyPaymentGateway(0, true);
        ExerciseSupport.expectEquals(
            "permanent-failure-stops-immediately",
            new RetryResult(null, 1, List.of(), "permanent-failure"),
            retryWithBackoff(4, 100, permanentFailureGateway::capture)
        );

        FlakyPaymentGateway exhaustedGateway = new FlakyPaymentGateway(5, false);
        ExerciseSupport.expectEquals(
            "retry-budget-exhausted",
            new RetryResult(null, 3, List.of(100L, 200L), "exhausted"),
            retryWithBackoff(3, 100, exhaustedGateway::capture)
        );
    }

    @FunctionalInterface
    interface ThrowingSupplier<T> {
        T get();
    }
}

File: labs/algorithms-live-coding/kotlin/src/main/kotlin/learning/livecoding/kotlin/RetryBackoffExercise.kt
package learning.livecoding.kotlin

object RetryBackoffExercise : Exercise {
    override val id = "retry-backoff"
    override val title = "Retry with Exponential Backoff"
    override val summary =
        "Retry only transient failures, increase the delay each time, and stop once the call succeeds or the retry budget is exhausted."

    data class RetryResult(
        val value: String?,
        val attempts: Int,
        val scheduledBackoffsMs: List<Long>,
        val status: String,
    )

    private class TransientFailure(message: String) : RuntimeException(message)

    private class PermanentFailure(message: String) : RuntimeException(message)

    private class FlakyPaymentGateway(
        private val transientFailuresBeforeSuccess: Int,
        private val permanentFailure: Boolean = false,
    ) {
        var calls: Int = 0
            private set

        fun capture(): String {
            calls++
            if (permanentFailure) throw PermanentFailure("card-declined")
            if (calls <= transientFailuresBeforeSuccess) throw TransientFailure("timeout")
            return "captured"
        }
    }

    private fun retryWithBackoff(
        maxAttempts: Int,
        initialBackoffMs: Long,
        operation: () -> String,
    ): RetryResult {
        var attempts = 0
        var nextBackoffMs = initialBackoffMs
        val scheduledBackoffs = mutableListOf<Long>()

        while (attempts < maxAttempts) {
            attempts++
            try {
                return RetryResult(
                    value = operation(),
                    attempts = attempts,
                    scheduledBackoffsMs = scheduledBackoffs,
                    status = "success",
                )
            } catch (_: PermanentFailure) {
                return RetryResult(
                    value = null,
                    attempts = attempts,
                    scheduledBackoffsMs = scheduledBackoffs,
                    status = "permanent-failure",
                )
            } catch (_: TransientFailure) {
                if (attempts == maxAttempts) {
                    return RetryResult(
                        value = null,
                        attempts = attempts,
                        scheduledBackoffsMs = scheduledBackoffs,
                        status = "exhausted",
                    )
                }
                scheduledBackoffs += nextBackoffMs
                nextBackoffMs *= 2
            }
        }

        error("Retry loop should have returned before reaching this point")
    }

    override fun run() {
        val successGateway = FlakyPaymentGateway(transientFailuresBeforeSuccess = 2)
        ExerciseSupport.expectEquals(
            "eventual-success",
            RetryResult("captured", attempts = 3, scheduledBackoffsMs = listOf(100L, 200L), status = "success"),
            retryWithBackoff(maxAttempts = 4, initialBackoffMs = 100, operation = successGateway::capture),
        )

        val permanentFailureGateway = FlakyPaymentGateway(transientFailuresBeforeSuccess = 0, permanentFailure = true)
        ExerciseSupport.expectEquals(
            "permanent-failure-stops-immediately",
            RetryResult(null, attempts = 1, scheduledBackoffsMs = emptyList(), status = "permanent-failure"),
            retryWithBackoff(maxAttempts = 4, initialBackoffMs = 100, operation = permanentFailureGateway::capture),
        )

        val exhaustedGateway = FlakyPaymentGateway(transientFailuresBeforeSuccess = 5)
        ExerciseSupport.expectEquals(
            "retry-budget-exhausted",
            RetryResult(null, attempts = 3, scheduledBackoffsMs = listOf(100L, 200L), status = "exhausted"),
            retryWithBackoff(maxAttempts = 3, initialBackoffMs = 100, operation = exhaustedGateway::capture),
        )
    }
}
